internal/dot/graph.go

```diff
@@ -252,6 +270,50 @@ func (dg *Graph) addToGroup(r *Result, id CtorID) {
 	group.Results = append(group.Results, r)
 }
 
+// Prune removes elements from the graph that do not have failed results.
+// Removing elements that do not have failing results makes the graph easier to debug, since non-failing
+// nodes and edges can clutter the graph and don't help the user debug.
+func (dg *Graph) Prune() {
+	failedResults := make(map[*Result]bool)
+	for _, v := range dg.Failed.RootCauses {
+		failedResults[v] = true
+	}
+	for _, v := range dg.Failed.TransitiveFailures {
+		failedResults[v] = true
+	}
+
+	dg.pruneCtors(failedResults)
+	dg.pruneGroups(failedResults)
```

Comment:
I think if you do the book keeping in advance, this gets simplified a bit.
You can probably just maintain a list of failed groupKeys and CtorIDs on the
FailedNodes object as FailNodes and FailGroupNodes methods are called. Then in
Prune, it's just a matter of iterating through those lists once rather than
traversing down to look for failed Results.